editors/code/src/dependencies_provider.ts

```diff
@@ -0,0 +1,154 @@
+    }
+
+    private _onDidChangeTreeData: vscode.EventEmitter<Dependency | DependencyFile | undefined | null | void> = new vscode.EventEmitter<Dependency | undefined | null | void>();
+
+    readonly onDidChangeTreeData: vscode.Event<Dependency | DependencyFile | undefined | null | void> = this._onDidChangeTreeData.event;
+
+
+    getDependency(filePath: string): Dependency | DependencyFile | undefined {
+        return this.dependenciesMap[filePath.toLowerCase()];
+    }
+
+    contains(filePath: string): boolean {
+        return filePath.toLowerCase() in this.dependenciesMap;
+    }
+
+    refresh(): void {
+        this._onDidChangeTreeData.fire();
+    }
+
+    getParent?(element: Dependency | DependencyFile): vscode.ProviderResult<Dependency | DependencyFile> {
+        if (element instanceof Dependency) return undefined;
+        return element.parent;
+    }
+
+    getTreeItem(element: Dependency | DependencyFile): vscode.TreeItem | Thenable<vscode.TreeItem> {
+        if (element.id! in this.dependenciesMap) return this.dependenciesMap[element.id!];
+        return element;
+    }
+
+    getChildren(element?: Dependency | DependencyFile): vscode.ProviderResult<Dependency[] | DependencyFile[]> {
+        return new Promise((resolve, _reject) => {
+            if (!this.workspaceRoot) {
+                void vscode.window.showInformationMessage('No dependency in empty workspace');
+                return Promise.resolve([]);
+            }
+
+            if (element) {
+                const files = fs.readdirSync(element.dependencyPath).map(fileName => {
+                    const filePath = fspath.join(element.dependencyPath, fileName);
+                    const collapsibleState = fs.lstatSync(filePath).isDirectory() ?
+                        vscode.TreeItemCollapsibleState.Collapsed :
+                        vscode.TreeItemCollapsibleState.None;
+                    const dep = new DependencyFile(
+                        fileName,
+                        filePath,
+                        element,
+                        collapsibleState
+                    );
+                    this.dependenciesMap[dep.dependencyPath.toLowerCase()] = dep;
+                    return dep;
+                });
+                return resolve(
+                    files
+                );
+            } else {
+                return resolve(this.getRootDependencies());
+            }
+        });
+    }
+
+    private async getRootDependencies(): Promise<Dependency[]> {
+        const registryDir = fspath.join(os.homedir(), '.cargo', 'registry', 'src');
+        const basePath = fspath.join(registryDir, fs.readdirSync(registryDir)[0]);
+        const deps = await this.getDepsInCartoTree(basePath);
+        const stdlib = await this.getStdLib();
+        this.dependenciesMap[stdlib.dependencyPath.toLowerCase()] = stdlib;
+        return [stdlib].concat(deps);
+    }
+
+    private async getStdLib(): Promise<Dependency> {
```

Comment:
Sure, can you please point me out on where can i find this logic in the server? ty

Reply:
There's https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide/src/view_crate_graph.rs for "View crate graph" and https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/project_model/src/workspace.rs#L140 takes care of loading a workspace. The sysroot is detected by `Sysroot::load` via `Sysroot::discover`.

Reply:
> This duplicates the logic in the server. I think we should do as much as possible there instead of in the client code.

I'm new here, but I disagree.  The server-side implementation takes some of the burden off of the developers of non-VSCode clients.

Reply:
I as well think it should be on the server side.  I'm sorry if I misunderstood something or if I wasn't clear.  Thanks.